File: src/arduinobot_firmware/arduinobot_firmware/simple_serial_transmitter.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import serial

class SimpleSerialTransmitter(Node):
    def __init__(self):
        super().__init__('simple_serial_transmitter')
        self.declare_parameter("port", "/dev/ttyACM0")
        self.declare_parameter("baud_rate", 115200)

        self.port = self.get_parameter("port").value
        self.baud_rate = self.get_parameter("baud_rate").value

        self.subcriber_ = self.create_subscription(String,'serial_transmitter', self.onMessageReceived, 10)
        self.arduino_ = serial.Serial(port=self.port,baudrate=self.baud_rate, timeout=0.1)

        self.get_logger().info('Subscribed to serial_transmitter')

    def onMessageReceived(self, msg):
        self.get_logger().info(f'Received message: {msg.data}')
        self.arduino_.write(msg.data.encode("utf-8"))
        self.get_logger().info(f'Message forwarded to robot')

def main(args=None):
    rclpy.init()
    simpleSerialTransmitter = SimpleSerialTransmitter()
    rclpy.spin(simpleSerialTransmitter)
    simpleSerialTransmitter.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/arduinobot_firmware/src/simple_serial_transmitter.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>
#include <libserial/SerialPort.h>

using std::placeholders::_1;

class SimpleSerialTransmitter : public rclcpp::Node
{
    public:
        SimpleSerialTransmitter() : Node("simple_serial_transmitter")
        {
            declare_parameter<std::string>("port","/dev/ttyACM0");
            std::string port = get_parameter("port").as_string();

            subscriber_ = create_subscription<std_msgs::msg::String>("serial_transmitter", 10, 
                std::bind(&SimpleSerialTransmitter::OnMessageReceived, this, _1));

            arduino_.Open(port);
            arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);

            RCLCPP_INFO(this->get_logger(), "CPP Arduino Transmitter ready");
        }

        ~SimpleSerialTransmitter(){
            arduino_.Close();
        }

        void OnMessageReceived(const std_msgs::msg::String::SharedPtr msg) 
        {
            RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg->data.c_str());
            arduino_.Write(msg->data);
            RCLCPP_INFO(this->get_logger(), "Message forwarded to robot");
        }

    private:
        rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;
        LibSerial::SerialPort arduino_;

};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimpleSerialTransmitter>());
    rclcpp::shutdown();
    return 0;
}